kotlin/viewregistry-android/src/main/java/com/squareup/viewregistry/modal/MainAndModalContainer.kt

```diff
@@ -0,0 +1,217 @@
+ * to displayed views that implement that interface themselves.
+ */
+class MainAndModalContainer(
+  context: Context,
+  attributeSet: AttributeSet?
+) : FrameLayout(context, attributeSet), HandlesBack {
+  constructor(context: Context) : this(context, null)
+
+  private val body: View? get() = getChildAt(0)
+  private var dialogs: List<DialogRef> = emptyList()
+  private val takeScreensSubs = CompositeDisposable()
+
+  private val attached = BehaviorSubject.create<Boolean>()
+
+  // Kind of a hack: use a Coordinator to create observable attached property.
+  // TODO should probably just drop Coordinators for RxBinding.
+  init {
+    Coordinators.bind(this) {
+      object : Coordinator() {
+        override fun attach(view: View) {
+          attached.onNext(true)
+        }
+
+        override fun detach(view: View) {
+          attached.onNext(false)
+
+          // TODO(https://github.com/square/workflow/issues/51)
+          // Not good enough, the stupid Activity cleans it up and shames us about "leaks" in logcat
+          // before this point. Try to use a lifecycle observer to clean that up.
+          dialogs.forEach { it.dialog.hide() }
+          dialogs = emptyList()
+        }
+      }
+    }
+  }
+
+  override fun onBackPressed(): Boolean {
+    // This should only be hit if there are no dialogs showing, so we only
+    // need to consider the body.
+    return (body as? HandlesBack)?.onBackPressed() == true
+  }
+
+  fun takeScreens(
+    screens: Observable<out MainAndModalScreen<*, *>>,
+    viewRegistry: ViewRegistry
+  ) {
+    takeScreensSubs.clear()
+
+    // This looks like we're leaking subscriptions to screens, since we only unsubscribe
+    // if takeScreens() is called again. It's fine, though:  the switchMap in whileAttached()
+    // ensures that we're only subscribed to it while this view is attached to a window.
+
+    // Create a new body view each time the type of [MainAndModalScreen.main] changes.
+    takeScreensSubs.add(screens
+        .whileAttached()
+        .distinctUntilChanged { screen -> screen.main::class }
+        .subscribe {
+          removeAllViews()
+          addView(it.viewForMain(screens, viewRegistry, this))
+        }
+    )
+
+    // Compare the new modals list to the set of dialogs already on display. Show
+    // any new ones, throw away any stale ones.
+    takeScreensSubs.add(screens
+        .whileAttached()
+        .subscribe { mainAndModal ->
+          val newDialogs = mutableListOf<DialogRef>()
+          for ((i, modalScreen) in mainAndModal.modals.withIndex()) {
+            newDialogs +=
+              if (dialogs.size < i && dialogs[i].screen::class == modalScreen::class) {
+                dialogs[i]
+              } else {
+                DialogRef(modalScreen, mainAndModal.showDialog(i, screens, viewRegistry, this))
+              }
+          }
+
+          dialogs.filter { it !in newDialogs }
```

Comment:
Nit: Could also do `(dialogs - newDialogs)` I believe, slightly more concise and I think just as expressive. Also, the kotlin runtime actually tries to optimize it a little if you use minus (TIL myself).

Here's the optimizations, if you're curious:
![image](https://user-images.githubusercontent.com/101754/52750365-df59fc00-2fa0-11e9-91f7-81f6b4ee1222.png)
![image](https://user-images.githubusercontent.com/101754/52750372-e8e36400-2fa0-11e9-8f18-c4a2eb428c01.png)

Reply:
Internal-only slack link unfortunately, but [some interesting discussion of this technique ensued](https://square.slack.com/archives/C0RN3QDN0/p1550099293015000).

Reply:
I had no idea that was even an option!

```
          (dialogs - newDialogs).forEach { it.dialog.hide() }
          dialogs = newDialogs
```